src/workerd/api/streams/writable.c++

```diff
@@ -365,6 +365,154 @@ private:
   }
 };
 
+// In order to support JavaScript-backed WritableStreams that do not have a backing
+// WritableStreamSink, we need an alternative version of the WritableStreamRpcAdapter
+// that will arrange to acquire the isolate lock when necessary to perform writes
+// directly on the WritableStreamController. Note that this approach is necessarily
+// a lot slower
+class WritableStreamJsRpcAdapter final: public capnp::ExplicitEndOutputStream {
+public:
+  WritableStreamJsRpcAdapter(IoContext& context, jsg::Ref<WritableStreamDefaultWriter> writer)
+      : context(context), writer(kj::mv(writer)) {}
+
+  ~WritableStreamJsRpcAdapter() noexcept(false) {
+    weakRef->invalidate();
+    doneFulfiller->fulfill();
+  }
+
+  // Returns a promise that resolves when the stream is dropped. If the promise is canceled before
+  // that, the stream is revoked.
+  kj::Promise<void> waitForCompletionOrRevoke() {
+    auto paf = kj::newPromiseAndFulfiller<void>();
+    doneFulfiller = kj::mv(paf.fulfiller);
+
+    return paf.promise.attach(kj::defer([weakRef = weakRef->addRef()]() mutable {
+      KJ_IF_SOME(obj, weakRef->tryGet()) {
+        // Stream is still alive, revoke it.
+        auto exception = cancellationException();
+        if (!obj.canceler.isEmpty()) {
+          obj.canceler.cancel(kj::cp(exception));
+        }
+        KJ_DEFER(obj.writer = kj::none);
+        KJ_IF_SOME(writer, obj.writer) {
+          obj.context.addTask(obj.context.run(
+              [writer=kj::mv(writer), exception=kj::mv(exception)](Worker::Lock& lock) mutable {
+            jsg::Lock& js = lock;
+            auto ex = js.exceptionToJs(kj::mv(exception));
+            return IoContext::current().awaitJs(lock,
+                writer->abort(lock, ex.getHandle(js)));
+          }));
+        }
+      }
+    }));
+  }
+
+  kj::Promise<void> write(const void* buffer, size_t size) override {
+    if (size == 0) return kj::READY_NOW;
+    return canceler.wrap(context.run(
+        [ref=weakRef->addRef(), buffer, size](Worker::Lock& lock) mutable
+            -> kj::Promise<void> {
+      jsg::Lock& js = lock;
+      auto& context = IoContext::current();
+      KJ_IF_SOME(obj, ref->tryGet()) {
```

Comment:
`tryGet()` here can't possibly return null because the caller of `write()` must cancel the promise before allowing the stream to be destroyed. So we know the stream is still valid here (or we already have a UAF anyway).

So you can just capture `this` rather than capturing `weakRef->addRef()` above.

Same for other methods below.